### PR TITLE
ref(replay): More efficient deserialization

### PR DESCRIPTION
**Pull request by jjbayer**
After deploying https://github.com/getsentry/relay/pull/1678, we saw a rise in memory consumption. We narrowed down the reason to deserialization of replay recordings, so this PR attempts to replace those deserializers with more efficient versions that do not parse an entire `serde_json::Value` to get the tag (`type`, `source`) of the enum.

A custom deserializer is necessary because serde does not support [integer tags for internally tagged enums](https://github.com/serde-rs/serde/issues/745).

- [x] Custom deserializer for `NodeVariant`, based on serde's own `derive(Deserialize)` of internally tagged enums.
- [x] Custom deserializer for `recording::Event`, based on serde's own `derive(Deserialize)` of internally tagged enums.
- [x] Custom deserializer for `IncrementalSourceDataVariant`, based on serde's own `derive(Deserialize)` of internally tagged enums.
- [x] Box all enum variants.

### Benchmark comparison

Ran a criterion benchmark on `rrweb.json`. It does not tell us anything about memory consumption, but the reduced cpu usage points to simpler deserialization:

#### Before

```
rrweb/1                 time:   [142.37 ms 148.17 ms 155.61 ms]
```

#### After

```
rrweb/1                 time:   [31.474 ms 31.801 ms 32.137 ms]
```

#skip-changelog

### CODE REVIEW

File: relay-replays/src/recording.rs
Comment (jjbayer) [inline relay-replays/src/recording.rs]: The basic idea here is to use serde's `Content` for more efficient deserialization, which is what serde itself uses for [deserializing internally tagged enums](https://cs.github.com/serde-rs/serde/blob/b7be637e8cf73dcbea28c6ea960604371726171d/serde_derive/src/de.rs?q=repo%3Aserde-rs%2Fserde+ContentDeserializer#L1332).

Since `Content` is not public API, we might also try using [`serde_json::RawValue`](https://docs.rs/serde_json/latest/serde_json/value/struct.RawValue.html) instead.

File: relay-replays/src/recording.rs
Comment (jjbayer) [inline relay-replays/src/recording.rs]: @cmanallen what confused me in the existing impl: Some event types have timestamp `u64`, some have `f64`. Shouldn't they all have the same timestamp format?

File: relay-replays/src/recording.rs
Comment (mitsuhiko) [reply relay-replays/src/recording.rs]: Content internally has a secret interface to avoid re-buffering when used recursively. RawValue doesn’t habe that *and* it has the added issue that it uses in-band signalling which would also change the input interface of any incoming json payload.

File: relay-replays/src/recording.rs
Comment (jjbayer) [inline relay-replays/src/recording.rs:245]: Boxing variants reduces the size of the enum itself -> should help against memory fragmentation.

File: relay-replays/src/serialization.rs
Comment (jjbayer) [inline relay-replays/src/serialization.rs]: There might be better approaches than using a helper enum/struct here, but this was the simplest solution I could come up with.

File: relay-replays/src/serialization.rs
Comment (jjbayer) [inline relay-replays/src/serialization.rs]: This implementation was copy-pasted and adapted from the expanded `derive(Deserialize)` of a simple, internally tagged enum.

File: relay-replays/src/recording.rs
Comment (jjbayer) [inline relay-replays/src/recording.rs:362]: We should probably pick better names for these variants, but I'll leave that to a follow-up PR.

File: relay-replays/src/recording.rs
Comment (cmanallen) [reply relay-replays/src/recording.rs]: @jjbayer RRWeb timestamps are u64.  Sentry timestamps are f64.

File: relay-server/src/actors/processor.rs
Comment (jjbayer) [inline relay-server/src/actors/processor.rs:1096]: This was the change of the original PR. I cherry-picked it in to see the impact on a Canary deploy.

**Review by cmanallen — COMMENTED** [review]
This pull seems to solve the high memory issue (more time will tell but at this point it looks very promising).  How did this branch fix it?  Is heap allocated memory freed more easily than stack allocated?  Was it boxing the enum variants that fixed it or the other serialization improvements in `serialization.rs`?  In the future is there a way for me to know when to heap allocate vs stack allocate memory?  As always, thank you for your help.  Your expertise is appreciated.

edit: @jjbayer tagging you

File: relay-replays/Cargo.toml
Comment (jan-auer) [inline relay-replays/Cargo.toml:25]: We use criterion 0.3 everywhere else. To avoid duplicate dependencies, could you either use that version or update everything else?

File: relay-replays/benches/benchmarks.rs
Comment (jan-auer) [inline relay-replays/benches/benchmarks.rs]: _nit_: Combine imports.

File: relay-replays/src/recording.rs
Comment (jan-auer) [inline relay-replays/src/recording.rs]: This `pub(crate)` is not needed.

File: relay-replays/src/recording.rs
Comment (jan-auer) [inline relay-replays/src/recording.rs]: Please do not use enums as namespaces, this is a common pitfall even though it's safe in this specific case. We prohibit this in our style guide.

File: relay-replays/src/recording.rs
Comment (jan-auer) [inline relay-replays/src/recording.rs]: _nit_: Since you need these from the serialization module, you could either make `serialization` a submodule of `recording`, or move the schema to another crate module that's then not exported.

There's nothing wrong with pub(crate) per se, but I think the way we're structuring the protocol and serialization is a bit cumbersome.

File: relay-replays/src/recording.rs
Comment (jan-auer) [inline relay-replays/src/recording.rs]: Also for a follow-up: The enum should be declared after the structs it contains as per style guide.

File: relay-replays/src/serialization.rs
Comment (jan-auer) [inline relay-replays/src/serialization.rs]: _nit_: While it's common to put these enums and visitors inside the `deserialize` implementation in larger modules, here we have an entire file just for that. I'd move them into the module scope for readability (less indentation, clearer separation between main deserialize logic and helper structs).

Comment (jan-auer) [inline relay-replays/src/serialization.rs]: This is not a strong objection, but the `EventType` enum is probably redundant. You might as well deserialize a `u64` directly and then match on the tag of the tagged content directly, raising the error in the match.

Note that below you did this for `IncrementalSourceDataVariant` (even though there's still dead code for `SourceVisitor`).

File: relay-replays/src/serialization.rs
Comment (jan-auer) [inline relay-replays/src/serialization.rs]: If you keep `EventType`, you should call `deserialize_u64` here directly, since this is the only type you expect. In theory, this would also make the deserializer compatible with non-self-describing formats.

File: relay-replays/src/serialization.rs
Comment (jan-auer) [inline relay-replays/src/serialization.rs]: Since we're using a private serde API here against repeated explicit advice of the maintainer, we should leave a great explanation in a code comment why that is.

File: relay-replays/src/serialization.rs
Comment (jan-auer) [inline relay-replays/src/serialization.rs]: Did you have that code expand/generate somewhere? It reads like machine-written or macro code.

```suggestion
            EventType::T0 => Value::deserialize(content_deserializer).map(Event::T0),
```

File: relay-replays/src/serialization.rs
Comment (jan-auer) [inline relay-replays/src/serialization.rs]: _nit_: If you introduce a `new` function for this, it will probably fit in one line and improve readability.

File: relay-replays/src/serialization.rs
Comment (jan-auer) [reply relay-replays/src/serialization.rs]: I'd stick with this, too. There's probably a way to get this working with remote serialization, but given that this type is rather static and won't change much, there's not much downside to the second enum.

File: relay-replays/src/serialization.rs
Comment (jan-auer) [inline relay-replays/src/serialization.rs]: The SourceVisitor is dead code since you deserialize with a `TaggedContentVisitor<u8>` directly.

**Review by jan-auer — CHANGES_REQUESTED** [review]

**Review by jan-auer — COMMENTED** [review]
One more general question: The parsed replay format uses a lot of `serde_json::Value` and `String` where we could have boolean flags or custom enums. Could we replace those with the intended types? I'm pretty sure that allows us to further reduce memory utilization and speed up parsing (though that would need to be verified). Potentially we could even revert some of the boxing because of that (though we'll need to keep it where there's large size differences between variants).

It'll also contribute to a stricter schema that should be easier to work with in the later parts of the pipeline.

Obviously, this would be a functional change with the potential to introduce regressions, so I'm happy for that to go in a follow-up PR. cc @cmanallen

**Comment by jan-auer** [comment]
> How did this branch fix it? Is heap allocated memory freed more easily than stack allocated?

@cmanallen There are actually two separate fixes in this PR:

**1. The size of enums.** 

An enum is a discriminator + one of the values. That means, the static size of this type is the size of its largest variant + usually one byte. If you now have a variant that (recursively) contains a lot of data, this blows up the size of your enum even if all the other variants are tiny. 

For example, the `Node` was 248 bytes large, regardless of whether all that data was needed or not. This always had to be allocated. And since Node is a recursive tree type, this quickly accumulates to a lot of memory needed. 

Joris' change reduces the size of Node to exactly one pointer size (8 bytes) by moving data to the heap. However, that heap allocation can now be smaller depending on which exact variant is being allocated. Overall, this means less wasteful memory usage and fewer empty space in between.

**2. Recursive Parsing**

The `Deserialize` impl for `Node` first parsed the _entire_ tree into a `Value`. Then you match on the top level and recurse. This means at every level of parsing a Node, you parse the tree recursively. This lead to exponential behavior wrt the depth of the tree. With Joris' change, we now only deserialize the top level.

Sadly, serde doesn't make this straight-forward. There's a built-in derive for enums with "string" tags (discriminator fields), but not if those fields are numbers. We now need to use serde's internal APIs to accomplish the same.

File: relay-replays/src/recording.rs
Comment (jjbayer) [reply relay-replays/src/recording.rs]: Agreed.

**Comment by cmanallen** [comment]
@jan-auer 

> Could we replace those with the intended types? [...] It'll also contribute to a stricter schema that should be easier to work with in the later parts of the pipeline.

I can make those optimizations in another PR.  My goal is to have a strict schema but unfortunately the RRWeb library has quirks which can lead to deserialization errors when I don't account for the full spectrum of types a field can occupy.  I'd like to be more permissive with `Value` for now and introduce stricter typing as we understand the full range of payload variants.

**Comment by cmanallen** [comment]
@jan-auer 

> Joris' change reduces the size of Node to exactly one pointer size (8 bytes) by moving data to the heap. However, that heap allocation can now be smaller depending on which exact variant is being allocated. Overall, this means less wasteful memory usage and fewer empty space in between.

👍 Great explanation, thanks!

File: relay-replays/src/recording.rs
Comment (jjbayer) [reply relay-replays/src/recording.rs]: You're right, `serialization` should be a submodule of `recording`.

File: relay-replays/src/serialization.rs
Comment (jjbayer) [reply relay-replays/src/serialization.rs]: Yes, I basically copied the `cargo expand` of an internally tagged enum and tweaked it to work with integer tags. That is also where the `EventType` indirection comes from. Will simplify.

File: relay-replays/Cargo.toml
Comment (jjbayer) [reply relay-replays/Cargo.toml:25]: Updated to 0.4 everywhere.

**Review by jan-auer — APPROVED** [review]

**Comment by mitsuhiko** [comment]
I'm surprised and quite a bit disappointed that clippy did not scream that the enum is huge. I wonder if this is something that is something that we can report upstream.

**Comment by jjbayer** [comment]
> I'm surprised and quite a bit disappointed that clippy did not scream that the enum is huge. I wonder if this is something that is something that we can report upstream.

@mitsuhiko if I'm not mistaken, that clippy only checks large _differences_ between variant sizes: https://github.com/rust-lang/rust-clippy/blob/96c28d1f69a120de7fcdbc40fb17610a407a4900/clippy_lints/src/large_enum_variant.rs#L18-L19

**Pull request merged**
